<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2025-4</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Daniel Rojo Mata
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 2.  Graficación en Python e Itegración Numérica


### Programación básica para la física
### Alumno: Rivas Rodriguez Luis Enrique
---

### **EJERCICIOS**:

$\;$

#### **1. Altitud de un satélite**: se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos.

#### **(a)** Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es:

<h3 align="center">$h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R$,</h3> 

donde $G = 6,67 × 10^{−11} m^3 kg^{−1}s^{−2}$ es la constante gravitatoria de Newton, $M = 5,97 ×
1024 kg$ es la masa de la Tierra, y $R = 6371 km$ es su radio.




Para que un satélite se mantenga en un órbita circular debemos considerar que las fuerzas centrípeta debe ser igual a la fuerza gravitacional que lo mantiene cerca del planeta entonces tenemos que
$F_c=F_g$


Entonces usando la fórmula gravitacional de Newton donde
$F_g= \frac{G \cdot m_1 \cdot m_2}{r^2} $
y la de la fuerza centrípeta $\frac{mv^2}{r}$ por lo tanto obtenemos las siguientes igualdad

$F_g= \frac{G \cdot M \cdot m}{r^2} = \frac{mv^2}{r}$

Para le caso de un satélite que orbita fuera de la tierra debemos tomar en cuenta a r como $r=R+h$ con R el radio del planeta y h la altura que alcanza el satélite, además consideremos que $v=\frac{d}{T}$ donde d es la distancia por vuelta completa al planeta, considerando un movimiento circular podemos expresar $v=\frac{2\pi (R+h)}{T}$ entonces podemos sustituir esta expresión y despejar de la siguiente manera


 $\frac{GM}{(R+h)^2}=\frac{mv^2}{R+h}$ $\implies$ $\frac{GM}{R+h}=v^2=\left(\frac{2\pi (R+h)}{T}\right)^2=\frac{4\pi^2(R+h)^2}{T^2}\implies GMT^2=4\pi^2 (R+h)^3\implies (R+h)^3=\frac{GMT^2}{4\pi^2}\implies R+h=\left(\frac{GMT^2}{4\pi^2}\right)^{\frac{1}{3}}\implies h=\left(\frac{GMT^2}{4\pi^2}\right)^{\frac{1}{3}}-R$
 
 Obteniendo así la expresión esperada.


#### **(b)** Escribe un programa que le pida al usuario que ingrese el valor deseado de $T$ y luego calcule e imprima la altura correcta en metros.


In [1]:
import numpy as np
import math as mt
import pandas as pd


In [2]:
T=int(input("Ingresa el tiempo T en segundo que le toma al satélite para orbitar el planeta:"))
print("Usando la fórmula de la altitud de un satélite se puede determinar dicha altitud con el valor dado")
G=6.67e-11
M=5.97e24
R=6.371e6
h=((G*M*T**2)/(4 *mt.pi**2))**(1/3)-R
redondeo=round(h,2)
print("Dado el tiempo T proporcionado la altitud del satélite será", redondeo, "metros")

Ingresa el tiempo T en segundo que le toma al satélite para orbitar el planeta: 86400


Usando la fórmula de la altitud de un satélite se puede determinar dicha altitud con el valor dado
Dado el tiempo T proporcionado la altitud del satélite será 35855910.18 metros


#### **(c)** Utiliza tu programa para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada *órbita "geosincrónica"*), una vez cada **90 minutos** y una vez cada **45 minutos**. 
#### ¿Qué concluyes del último de estos cálculos?

In [3]:
Tm=90
ts=Tm*60
G=6.67e-11
M=5.97e24
R=6.371e6
h=((G*M*ts**2)/(4 *mt.pi**2))**(1/3)-R
redondeo=round(h,2)
print("Dado el tiempo T proporcionado la altitud del satélite será", redondeo, "metros")

Dado el tiempo T proporcionado la altitud del satélite será 279321.63 metros


In [4]:
Tm=45
ts=Tm*60
G=6.67e-11
M=5.97e24
R=6.371e6
h=((G*M*ts**2)/(4 *mt.pi**2))**(1/3)-R
redondeo=round(h,2)
print("Dado el tiempo T proporcionado la altitud del satélite será", redondeo, "metros")

Dado el tiempo T proporcionado la altitud del satélite será -2181559.9 metros


Como se observa en los resultados a una cierta cantidad de tiempo los resultados se vuelven negativos indicando que existe un mínimo de tiempo de órbita que está estrechamente relacionado con la altitud de dichos satélites, que al bajar de éste mínimo da resultados irreales o que indican que es físicamente imposible mantener un satélite en órbita, entonces, podemos concluir que los valores del tiempo de órbita y altitud son dependientes entre sí.

#### **2. Relatividad especial:** Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 

#### Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

#### **(a)** en el marco de reposo de un observador en la Tierra y

#### **(b)** como lo percibiría un pasajero a bordo de la nave. 

In [5]:
import numpy as np
import math as mt
c=3e8
x=float(input("Ingresa la distancia a recorrer en años luz:"))
v=float(input("Ingresa la velocidad expresada como una fracción de la velopcidad de la luz (c)"))
t_tierra=x/v
redondeot=round(t_tierra,2)
print("a) Para un observador en la tierra la nave llega en", redondeot, "años")
gamma=1/mt.sqrt(1-v**2)
t_nave=redondeot/gamma
redondeon=round(t_nave,2)
print("b) Mientras para la nave el viaje será de",redondeon,"años")

Ingresa la distancia a recorrer en años luz: 95
Ingresa la velocidad expresada como una fracción de la velopcidad de la luz (c) 0.74


a) Para un observador en la tierra la nave llega en 128.38 años
b) Mientras para la nave el viaje será de 86.35 años


#### Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

In [7]:
import numpy as np
import math as mt
c=3e8
x=float(input("Ingresa la distancia a recorrer en años luz:"))
v=float(input("Ingresa la velocidad expresada como una fracción de la velocidad de la luz (c)"))
t_tierra=x/v
redondeot=round(t_tierra,2)
print("Para un observador en la tierra la nave llega en", redondeot, "años")
gamma=1/mt.sqrt(1-v**2)
t_nave=redondeot/gamma
redondeon=round(t_nave,2)
print("Mientras para la nave el viaje será de",redondeon,"años")

Ingresa la distancia a recorrer en años luz: 10
Ingresa la velocidad expresada como una fracción de la velocidad de la luz (c) 0.99


Para un observador en la tierra la nave llega en 10.1 años
Mientras para la nave el viaje será de 1.42 años


#### **3. Órbitas planetarias**: En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su \textit{perihelio}), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

#### **(a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o \textit{afelio}, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3>
       
#### donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

#### Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.


Considerando de la ecuación cuadrática:

$$
v_2^2 - \frac{2GM}{v_1 \ell_1} v_2 - \left( v_1^2 - \frac{2GM}{\ell_1} \right) = 0
$$

Como comúnmente se sabe esta solo tiene dos soluciones reales, ya que el discriminante es positivo. Aplicando la fórmula cuadrática:

$$
v_2 = \frac{\dfrac{2GM}{v_1 \ell_1} \pm \sqrt{ \left( \dfrac{2GM}{v_1 \ell_1} \right)^2 + 4\left( v_1^2 - \dfrac{2GM}{\ell_1} \right) }}{2}
$$

Como se trata de sumar o restar una cantidad positiva, la raíz con el signo “–” es la más pequeña de las dos. Sin embargo, para justificar físicamente que esa raíz representa la verdadera velocidad $v_2$, aplicamos conservación del momento angular:

$$
\ell_1 v_1 = \ell_2 v_2 \quad \Rightarrow \quad v_2 = \frac{\ell_1 v_1}{\ell_2}
$$

Y dado que $\ell_2 > \ell_1$ (el afelio está más lejos que el perihelio en una órbita elíptica), se concluye que:

$$
\frac{\ell_1}{\ell_2} < 1 \quad \Rightarrow \quad v_2 < v_1
$$

Es decir, el valor válido de $v_2$ debe ser menor que $v_1$. Al mirar las dos raíces de la ecuación, sólo una cumple con esta condición siendo así la más pequeña.


#### **(b)**  Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center"><b>Semieje mayor</b>: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2),$</h3>

<h4 align="center"><b>Semieje menor</b>: $\; \; b = \sqrt{\ell_1\ell_2},$</h3>

<h4 align="center"><b>Período orbital</b>: $\; \; T = {2\pi ab\over\ell_1 v_1},$</h3>

<h4 align="center"><b>Excentricidad orbital</b>: $\; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1},$</h3>

#### Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades $\ell_2$, $v_2$, $T$ y $e$.


In [44]:
import math as mt
v1=float(input("Ingrese el valor de la velocidad en el perihelio v1"))
l1=float(input("Ingrese el valor de la distancia en le perihelio l1"))
G = 6.6743e-11
M = 1.989e30  
A = 1
B = - (2 * G * M) / (v1 * l1)
C = - (v1**2 - (2 * G * M / l1))
discriminante = B**2 - 4 * A * C
if discriminante < 0:
    print("El discriminante es negativo, el cálculo es inválido")
else:
    v2 = (-B - mt.sqrt(discriminante)) / (2 * A)
print("El valor de la velocidad en el afelio =", v2)
l2=(l1*v1)/v2
a=(1/2)*(l1+l2)
b=mt.sqrt(l1*l2)
if b<0:
    print("La expresión es negativa, la raíz es inválida")
else:
    T=(2*mt.pi*a*b)/(l1*v1)
e=(l2-l1)/(l2+l1)
print(f"La distancia en su afelio l2 = {l2:.2f}")
print(f"El periodo orbital T = {T:.2f}")
T_dias = T / (60*60*24)
print(f"O bien es T = {T_dias:.2f} días")
print(f"La excentricidad e = {e:.5f}")


Ingrese el valor de la velocidad en el perihelio v1 30290
Ingrese el valor de la distancia en le perihelio l1 1.471e11


El valor de la velocidad en el afelio = 29297.965326790112
La distancia en su afelio l2 = 152080833952.17
El periodo orbital T = 31551155.87
O bien es T = 365.18 días
La excentricidad e = 0.01665


#### **(c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [43]:
##Para la tierra
import math as mt
v1=3.0287e4
l1=1.4710e11
G = 6.6743e-11
M = 1.989e30  
A = 1
B = - (2 * G * M) / (v1 * l1)
C = - (v1**2 - (2 * G * M / l1))
discriminante = B**2 - 4 * A * C
if discriminante < 0:
    print("El discriminante es negativo, el cálculo es inválido")
else:
    v2 = (-B - mt.sqrt(discriminante)) / (2 * A)
print(f"El valor de la velocidad en el afelio = {v2:.2f}")
l2=(l1*v1)/v2
a=(1/2)*(l1+l2)
b=mt.sqrt(l1*l2)
if b<0:
    print("La expresión es negativa, la raíz es inválida")
else:
    T=(2*mt.pi*a*b)/(l1*v1)
e=(l2-l1)/(l2+l1)
print(f"La distancia en su afelio l2 = {l2:.2f}")
print(f"El periodo orbital T = {T:.2f}")
T_dias = T / (60*60*24)
print(f"O bien es T = {T_dias:.2f} días")
print(f"La excentricidad e = {e:.5f}")


El valor de la velocidad en el afelio = 29306.87
La distancia en su afelio l2 = 152019579575.61
El periodo orbital T = 31541466.67
O bien es T = 365.06 días
La excentricidad e = 0.01645


In [45]:
##Para el cometa
import math as mt
v1=5.4529e4
l1=8.7830e10
G = 6.6743e-11
M = 1.989e30  
A = 1
B = - (2 * G * M) / (v1 * l1)
C = - (v1**2 - (2 * G * M / l1))
discriminante = B**2 - 4 * A * C
if discriminante < 0:
    print("El discriminante es negativo, el cálculo es inválido")
else:
    v2 = (-B - mt.sqrt(discriminante)) / (2 * A)
print(f"El valor de la velocidad en el afelio = {v2:.2f}")
l2=(l1*v1)/v2
a=(1/2)*(l1+l2)
b=mt.sqrt(l1*l2)
if b<0:
    print("La expresión es negativa, la raíz es inválida")
else:
    T=(2*mt.pi*a*b)/(l1*v1)
e=(l2-l1)/(l2+l1)
print(f"La distancia en su afelio l2 = {l2:.2f}")
print(f"El periodo orbital T = {T:.2f}")
T_dias = T / (60*60*24)
print(f"O bien es T = {T_dias:.2f} días")
print(f"La excentricidad e = {e:.5f}")


El valor de la velocidad en el afelio = 908.05
La distancia en su afelio l2 = 5274268185967.95
El periodo orbital T = 2393959311.21
O bien es T = 27707.86 días
La excentricidad e = 0.96724


### **4.La fórmula semiempírica de la masa (FSM)**: En física nuclear, la fórmula de Weizsäcker (conocida tambien como fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un nucleo atómico; y está basada parcialmente en mediciones empíricas.

### En particular la fórmula se usa para calcular la energía de enlace nuclear aproximada  B, de un nucleo atómico con número atómico Z y número de masa A:

$B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A - 2Z)^2}{A} + \frac{a_5}{A^{1/2}}$

donde, en unidades de millones de electrón-volts:



$a_1 = 15.8, \quad a_2 = 18.3, \quad a_3 = 0.714, \quad a_4 = 23.2,$



y



$a_5 =  \begin{cases} 
0 & \text{si } A \text{ es impar}, \\
12.0 & \text{si } A \text{ y } Z \text{ son pares (ambos)}, \\
-12.0 & \text{si } A \text{ es par y } Z \text{ impar}.
\end{cases}$



### a) Escribe un programa que tome como entrada los valores de \( A \) y \( Z \), e imprima la energía de enlace \( B \) para el átomo correspondiente. Usa tu programa para encontrar la energía de enlace de un átomo con \( A = 58 \) y \( Z = 28 \). (Hint: La respuesta correcta es alrededor de los 490 MeV).

In [9]:
A=float(input("Ingrese el valor de masa A del átomo:"))
Z=float(input("Ingrese el número atómico de del elemento:"))
a1=15.8
a2=18.3
a3=0.714
a4=23.2
if A % 2 == 1:  # A impar
    a5 = 0
elif A % 2 == 0 and Z % 2 == 0:  # Ambos pares
    a5 = 12.0
elif A % 2 == 0 and Z % 2 == 1:  # A par y Z impar
    a5 = -12.0
B=a1*A-a2*A**(2/3)-a3*Z**2/A**(1/3)-a4*((A-2*Z)**2/A)+a5/A**(1/2)
print(f"La energía de enlace del átomo con A y Z proporcionados es: {B:.2f} MeV")

Ingrese el valor de masa A del átomo: 58
Ingrese el número atómico de del elemento: 28


La energía de enlace del átomo con A y Z proporcionados es: 497.56 MeV


### b) Modifica el programa del inciso anterior, para escribir una segunda versión que imprima no la energía de enlace total \( B \), sino la energía de unión por nucleón, que es \( B / A \).
    

In [12]:
A=float(input("Ingrese el valor de masa A del átomo:"))
Z=float(input("Ingrese el número atómico de del elemento:"))
a1=15.8
a2=18.3
a3=0.714
a4=23.2
if A % 2 == 1:  # A impar
    a5 = 0
elif A % 2 == 0 and Z % 2 == 0:  # Ambos pares
    a5 = 12.0
elif A % 2 == 0 and Z % 2 == 1:  # A par y Z impar
    a5 = -12.0
B=a1*A-a2*A**(2/3)-a3*Z**2/A**(1/3)-a4*((A-2*Z)**2/A)+a5/A**(1/2)
print(f"La energía de enlace del átomo con A y Z proporcionados es: {B:.2f} MeV")
EpN=B/A
print(f"La energía de unión por nucleón dado B y A es {EpN:.2f} MeV")

Ingrese el valor de masa A del átomo: 58
Ingrese el número atómico de del elemento: 28


La energía de enlace del átomo con A y Z proporcionados es: 497.56 MeV
La energía de unión por nucleón dado B y A es 8.58 MeV


### c) Escribe una tercera version del programa para que tome como entrada solo un valor del número atómico Z y luego pase por todos los valores de A desde A = Z hasta A = 3Z, para encontrar el que tiene la mayor energ ́ıa de enlace por nucleon. Este es el nucleo más estable con el número atómico dado. Haz que tu programa imprima el valor de A para este nucleo más estable y el valor de la energía de enlace por nucleon.

In [35]:
Z=int(input("Ingrese el número atómico de del elemento:"))
a1=15.8
a2=18.3
a3=0.714
a4=23.2
Mejor_A=Z
Mejor_EnergiaN=0
def EnergiaEnlace(A,Z):

    a1=15.8
    a2=18.3
    a3=0.714
    a4=23.2
    if A % 2 == 1:  # A impar
        a5 = 0
    elif A % 2 == 0 and Z % 2 == 0:  # Ambos pares
        a5 = 12.0
    elif A % 2 == 0 and Z % 2 == 1:  # A par y Z impar
        a5 = -12.0
    B=a1*A-a2*A**(2/3)-a3*Z**2/A**(1/3)-a4*((A-2*Z)**2/A)+a5/A**(1/2)
    return B
for A in range(Z,3*Z+1):
    B=EnergiaEnlace(A,Z)
    EnergiaNucleon=B/A
    if EnergiaNucleon > Mejor_EnergiaN:
        Mejor_EnergiaN=EnergiaNucleon
        Mejor_A=A
print(f"\nEl núcleo más estable para Z = {Z} es A = {Mejor_A}")
print(f"Energía de enlace por nucleón: {Mejor_EnergiaN:.2f} MeV")

Ingrese el número atómico de del elemento: 28



El núcleo más estable para Z = 28 es A = 62
Energía de enlace por nucleón: 8.70 MeV


### d) Finalmente, escribe una cuarta version del programa que, en lugar de tomar  ́ Z como entrada, se ejecute a traves de todos los valores de  ́ Z de 1 a 100 e imprima el valor mas estable de  ́ A para cada uno. ¿A que valor de  ́ Z se produce la energ ́ıa de enlace maxima por nucle  ́ on? (La respuesta correcta, en la vida real, es  ́ Z = 28, que corresponde al N ́ıquel).

In [3]:
a1=15.8
a2=18.3
a3=0.714
a4=23.2
def EnergiaEnlace(A,Z):

    a1=15.8
    a2=18.3
    a3=0.714
    a4=23.2
    if A % 2 == 1:  # A impar
        a5 = 0
    elif A % 2 == 0 and Z % 2 == 0:  # Ambos pares
        a5 = 12.0
    elif A % 2 == 0 and Z % 2 == 1:  # A par y Z impar
        a5 = -12.0
    B=a1*A-a2*A**(2/3)-a3*Z**2/A**(1/3)-a4*((A-2*Z)**2/A)+a5/A**(1/2)
    return B
Mejor_Z_Global = 0
Mejor_A_Global = 0
Mejor_Energia_Global = 0
for Z in range (1, 101):
    Mejor_A=Z
    Mejor_EnergiaN=0
    for A in range(Z,Z*3+1):
        B=EnergiaEnlace(A,Z)
        EnergiaNucleon=B/A
        if EnergiaNucleon > Mejor_EnergiaN:
            Mejor_EnergiaN=EnergiaNucleon
            Mejor_A=A
    print(f"Z = {Z:3} | Mejor A = {Mejor_A:3} | Energía por nucleón = {Mejor_EnergiaN:.2f} MeV")
     # Verificar si este Z tiene la mejor energía global
    if Mejor_EnergiaN > Mejor_Energia_Global:
        Mejor_Energia_Global = Mejor_EnergiaN
        Mejor_Z_Global = Z
        Mejor_A_Global = Mejor_A

print("\n" + "-" * 60)
print(f"Z con mayor energía por nucleón: Z = {Mejor_Z_Global}")
print(f"A más estable en ese Z: A = {Mejor_A_Global}")
print(f"Energía por nucleón máxima: {Mejor_Energia_Global:.2f} MeV")

Z =   1 | Mejor A =   3 | Energía por nucleón = 0.37 MeV
Z =   2 | Mejor A =   4 | Energía por nucleón = 5.32 MeV
Z =   3 | Mejor A =   7 | Energía por nucleón = 5.28 MeV
Z =   4 | Mejor A =   8 | Energía por nucleón = 6.47 MeV
Z =   5 | Mejor A =  11 | Energía por nucleón = 6.65 MeV
Z =   6 | Mejor A =  14 | Energía por nucleón = 7.20 MeV
Z =   7 | Mejor A =  15 | Energía por nucleón = 7.33 MeV
Z =   8 | Mejor A =  18 | Energía por nucleón = 7.72 MeV
Z =   9 | Mejor A =  19 | Energía por nucleón = 7.74 MeV
Z =  10 | Mejor A =  22 | Energía por nucleón = 8.04 MeV
Z =  11 | Mejor A =  25 | Energía por nucleón = 8.03 MeV
Z =  12 | Mejor A =  26 | Energía por nucleón = 8.24 MeV
Z =  13 | Mejor A =  29 | Energía por nucleón = 8.24 MeV
Z =  14 | Mejor A =  30 | Energía por nucleón = 8.38 MeV
Z =  15 | Mejor A =  33 | Energía por nucleón = 8.39 MeV
Z =  16 | Mejor A =  36 | Energía por nucleón = 8.49 MeV
Z =  17 | Mejor A =  37 | Energía por nucleón = 8.48 MeV
Z =  18 | Mejor A =  40 | Energ

### 5. La constante de Madelung  
En física de la materia condensada, la constante de Madelung da el potencial eléctrico  
total que siente un átomo en un sólido; y depende de las cargas de los otros átomos  
cercanos y de sus ubicaciones.  
Por ejemplo, el cristal de cloruro de sodio sólido (la sal de mesa), tiene átomos dispues-  
tos en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio  
una carga positiva $+e$ y los de cloro una negativa $-e$ (donde $e$ es la carga del electrón).  

Si etiquetamos cada posición en la red con tres coordenadas enteras $(i, j, k)$, entonces  
los átomos de sodio caen en posiciones donde $i + j + k$ es par, y los átomos de cloro en  
posiciones donde $i + j + k$ es impar.  
Consideremos un átomo de sodio en el origen, i.e. $i = j = k = 0$, y calculemos la  
constante de Madelung. Si el espaciado de los átomos en la red es $a$, entonces la distancia  
desde el origen al átomo en la posición $(i, j, k)$ es:  

$a \sqrt{i^2 + j^2 + k^2}$  

y el potencial en el origen creado por tal átomo es:  

$V(i, j, k) = \pm \dfrac{e}{4\pi\varepsilon_0 a \sqrt{i^2 + j^2 + k^2}}$  

siendo $\varepsilon_0$ la permitividad del vacío y el signo de la expresión se toma dependiendo de  
si $i + j + k$ es par o impar. Así entonces, el potencial total que siente el átomo de sodio  
es la suma de esta cantidad sobre todos los demás átomos. Supongamos una caja cú-  
bica alrededor del átomo de sodio en el origen, con $L$ átomos en todas las direcciones,  
entonces:  

$V_{\text{total}} = \sum\limits_{\substack{i,j,k=-L \\ (i,j,k)\ne (0,0,0)}}^{L} V(i, j, k) = \dfrac{e}{4\pi\varepsilon_0 a} \cdot M$  

donde $M$ es la constante de Madelung (al menos aproximadamente).  
Técnicamente, la constante de Madelung es el valor de $M$ cuando $L \to \infty$, pero se  
puede obtener una buena aproximación simplemente usando un valor grande de $L$.  
Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro  
de sodio. Utiliza un valor de $L$ tan grande como puedas, sin dejar que tu programa se  
ejecute en un tiempo razonable (un minuto o menos).


In [17]:
import time

L = int(input("Introduce el valor de L como la distancia desde el origen: "))
tiempo_maximo = 60 
inicio = time.time()

M = 0.0
completo = True

for i in range(-L, L + 1):
    for j in range(-L, L + 1):
        for k in range(-L, L + 1):
            if i == 0 and j == 0 and k == 0:
                continue
            r = (i**2 + j**2 + k**2)**0.5
            signo = 1 if (i + j + k) % 2 == 0 else -1
            M = M + signo / r

            if time.time() - inicio > tiempo_maximo:
                completo = False
                break
        if not completo:
            break
    if not completo:
        break

if completo:
    print(f" La aproximación para la de Madelung para L = {L} será {M:.6f}")
else:
    print("El tiempo se agotó. El valor de L =", L," tomó demasiado tiempo, el programa se cierra.")
    print("Prueba con un L más pequeño.")


Introduce el valor de L como la distancia desde el origen:  100


 La aproximación para la de Madelung para L = 100 será -1.741820


### 6. Coeficientes binomiales

El coeficiente binomial  
$ \binom{n}{k} $  
es un número entero igual a:

$ \binom{n}{k} = \dfrac{n!}{k!(n - k)!} = \dfrac{n \cdot (n - 1) \cdot (n - 2) \cdots (n - k + 1)}{1 \cdot 2 \cdots k} $

donde $k \geq 1$, o bien  
$ \binom{n}{0} = 1 $ cuando $k = 0$.


a) Utiliza esta fórmula para escribir una función llamada `binomial(n, k)` (o como  
tú quieras) que calcule el coeficiente binomial para un $n$ y $k$ dados.  
Asegúrate de que tu función devuelva la respuesta en forma de un número entero  
(no flotante) y proporcione el valor correcto de $1$ para el caso en que $k = 0$.

In [25]:
def binomial(n,k):
    if k==0:
        return 1
    numerador = 1
    denominador = 1
    for i in range (1, k+1):
        numerador = numerador*n-(i-1)
        denominador = denominador*i
    return numerador // denominador
n=int(input("Ingrese el numéro n:"))
k=int(input("Ingrese el numéro k:"))
print("El binomial de n y k dados es", binomial(n,k))

Ingrese el numéro n: 9
Ingrese el numéro k: 3


El binomial de n y k dados es 119


b) Usando tu función, escribe un programa que imprima las primeras 20 líneas del  
“triángulo de Pascal’’. La $n$-ésima línea del triángulo de Pascal contiene  
$n + 1$ números, que son los coeficientes  
$ \binom{n}{0}, \binom{n}{1}, \ldots, \binom{n}{n} $.  

De tal manera que las primeras líneas son:  
1  
1 1  
1 2 1  
1 3 3 1  
1 4 6 4 1


In [46]:
f=20
def binomial(n,k):
    if k==0:
        return 1
    numerador = 1
    denominador = 1
    for i in range (1, k+1):
        numerador = (numerador)*(n-(i-1))
        denominador = denominador*i
    return numerador // denominador
for n in range(f):
    linea=' '
    for k in range (n+1):
        linea=linea + str(binomial(n,k))+ ' '
    print(linea.strip())

1
1 1
1 2 1
1 3 3 1
1 4 6 4 1
1 5 10 10 5 1
1 6 15 20 15 6 1
1 7 21 35 35 21 7 1
1 8 28 56 70 56 28 8 1
1 9 36 84 126 126 84 36 9 1
1 10 45 120 210 252 210 120 45 10 1
1 11 55 165 330 462 462 330 165 55 11 1
1 12 66 220 495 792 924 792 495 220 66 12 1
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1


c) La probabilidad de que para una moneda no sesgada, lanzada $n$ veces, salga águila $k$ veces es:

$ p(k|n) = \dfrac{\binom{n}{k}}{2^n} $

Escribe un programa para calcular:  
1) La probabilidad total de que una moneda lanzada 100 veces, salga águila exactamente 60 veces.  
2) La probabilidad de que salga águila 60 veces o más.


In [47]:
def binomial(n, k):
    if k == 0 or k == n:
        return 1
    numerador = 1
    denominador = 1
    for i in range(1, k + 1):
        numerador =numerador* (n - (i - 1))
        denominador =denominador * i
    return numerador // denominador

def probabilidad(k, n):
    return binomial(n, k) / (2 ** n)


exct = probabilidad(60, 100)
print(f"La probabilidad exacta de 60 águilas en 100 lanzamientos será de: {exct:.10f}")

suma = 0
for k in range(60, 101):
    suma =suma + probabilidad(k, 100)
print(f"La probabilidad de 60 o más águilas en 100 lanzamientos será de: {suma:.10f}")


La probabilidad exacta de 60 águilas en 100 lanzamientos será de: 0.0108438667
La probabilidad de 60 o más águilas en 100 lanzamientos será de: 0.0284439668


### 7. Recursión

Una característica muy útil de las funciones que podemos definir es la *recursividad*, es decir, la capacidad de una función de llamarse a sí misma. Por ejemplo, considere la siguiente definición del factorial $n!$ de un entero positivo $n$:

$$
n! = 
\begin{cases}
1 & \text{si } n = 1, \\
n \times (n - 1)! & \text{si } n > 1.
\end{cases}
$$

Esto constituye una definición completa del factorial que nos permite calcular el valor de $n!$ para cualquier entero positivo. Podemos emplear esta definición directamente para crear una función para calcular factoriales, como esta (en Python):

```python
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)



Nota cómo si n no es igual a 1, la función se llama a sí misma para calcular el factorial de n - 1, de tal manera que al pedir print(factorial(5)), el intérprete de Python imprimirá correctamente la respuesta 120.

a) Anteriormente, en clase encontramos los números de Catalán $C_n$. Si hacemos una pequeña modificación, la definición de $C_n$ se puede reescribir de la siguiente forma:

$$
C_n =
\begin{cases}
1 & \text{si } n = 0, \\
\frac{4n - 2}{n + 1} C_{n-1} & \text{si } n > 0.
\end{cases}
$$

Escribe una función, usando recursividad, que calcule $C_n$ y utilízala para calcular e imprimir $C_{100}$.


In [57]:
def catalan(n):
        if n==0:
            Cn = 1
        else:
            Cn = ((4 * n - 2) * catalan(n - 1)) // (n + 1)
        return Cn
for i in range (0,101):
    print("C_",i,"=",catalan(i))

C_ 0 = 1
C_ 1 = 1
C_ 2 = 2
C_ 3 = 5
C_ 4 = 14
C_ 5 = 42
C_ 6 = 132
C_ 7 = 429
C_ 8 = 1430
C_ 9 = 4862
C_ 10 = 16796
C_ 11 = 58786
C_ 12 = 208012
C_ 13 = 742900
C_ 14 = 2674440
C_ 15 = 9694845
C_ 16 = 35357670
C_ 17 = 129644790
C_ 18 = 477638700
C_ 19 = 1767263190
C_ 20 = 6564120420
C_ 21 = 24466267020
C_ 22 = 91482563640
C_ 23 = 343059613650
C_ 24 = 1289904147324
C_ 25 = 4861946401452
C_ 26 = 18367353072152
C_ 27 = 69533550916004
C_ 28 = 263747951750360
C_ 29 = 1002242216651368
C_ 30 = 3814986502092304
C_ 31 = 14544636039226909
C_ 32 = 55534064877048198
C_ 33 = 212336130412243110
C_ 34 = 812944042149730764
C_ 35 = 3116285494907301262
C_ 36 = 11959798385860453492
C_ 37 = 45950804324621742364
C_ 38 = 176733862787006701400
C_ 39 = 680425371729975800390
C_ 40 = 2622127042276492108820
C_ 41 = 10113918591637898134020
C_ 42 = 39044429911904443959240
C_ 43 = 150853479205085351660700
C_ 44 = 583300119592996693088040
C_ 45 = 2257117854077248073253720
C_ 46 = 8740328711533173390046320
C_ 47 = 338

b) Euclides demostró que el máximo común divisor $g(m, n)$ de dos enteros no negativos $m$ y $n$ satisface que:

$$
g(m, n) =
\begin{cases}
m & \text{si } n = 0, \\
g(n, m \bmod n) & \text{si } n > 0.
\end{cases}
$$

Usando esta fórmula, escribe una función $g(m, n)$ que emplee recursividad para calcular el máximo común divisor de $m$ y $n$. Usa tu función para calcular e imprimir el máximo común divisor de $108$ y $192$.

In [61]:
def g(m,n):
    if n==0:
        return m
    else:
        return g(n, m % n)
n=int(input("Ingrese el número m:"))
m=int(input("Ingrese el número n:"))
print("El máximo común divisor emtre m y n dados es", g(m,n))

Ingrese el número m: 192
Ingrese el número n: 108


El máximo común divisor emtre m y n dados es 12


Comparando el cálculo de los números de Catalán del inciso 7a con lo hecho en clase, vemos que es posible hacer el cálculo de dos maneras: ya sea directamente o usando recursividad.

En la mayoría de los casos, si una cantidad se puede calcular sin recursividad, será más rápido y normalmente se recomienda hacerlo así, si es posible. Sin embargo, hay algunos cálculos que son esencialmente imposibles (o al menos mucho más difíciles) sin recursividad.